### DISCUSSION TITLE
Name must contain only ASCII letters, digits, and hyphens. (Parameter 'name')

### DISCUSSION THREAD

**Discussion by Depechie** [microsoft/aspire#2823] [Q&A] [answered]
Preview 4 of Aspire gives following exception when trying to start the AppHost:

```
Unhandled exception. System.ArgumentException: Resource name 'basket.api' is invalid. Name must contain only ASCII letters, digits, and hyphens. (Parameter 'name')
   at Aspire.Hosting.ApplicationModel.ModelName.ValidateName(String target, String name) in /_/src/Aspire.Hosting/ApplicationModel/ModelName.cs:line 39
   at Aspire.Hosting.ApplicationModel.Resource..ctor(String name) in /_/src/Aspire.Hosting/ApplicationModel/Resource.cs:line 30
   at Aspire.Hosting.ApplicationModel.ProjectResource..ctor(String name) in /_/src/Aspire.Hosting/ApplicationModel/ProjectResource.cs:line 10
   at Aspire.Hosting.ProjectResourceBuilderExtensions.AddProject[TProject](IDistributedApplicationBuilder builder, String name) in /_/src/Aspire.Hosting/Extensions/ProjectResourceBuilderExtensions.cs:line 27
```

The culprit is in defining a project like:

```
var basketAPI = builder.AddProject<Projects.otel_Basket_API>("basket.api")
```

This works fine with preview 3.

**Comment by davidfowl** [comment, marked as answer]
Yep, we changed the naming constraints. Change the . to a dash or underscore

**Reply by paulomorgado** [reply]
Why?

That will force Aspire into applications instead of helping.

What happens if I one can't control the service configuration?

**Reply by davidfowl** [reply]
These names often have to be valid uris as they are used for service discovery. So we based the naming rules loosely on the uri naming rules.

> That will force Aspire into applications instead of helping.

I don’t understand this. Can you clarify?

**Reply by paulomorgado** [reply]
If you're adding a container or a project that needs to have dots in the service name, this breaks that.

Changing those might not be possible, due to their semantic value for other tools.

If this constraint is being imposed, there should be a way (or guidance) on how to add a reference with a different name.

**Reply by davidfowl** [reply]
Do you have a real example of this?

**Reply by paulomorgado** [reply]
I have several services registered as `<company>.<system>.<service>`.

These are perfectly reasonable names and, just by registering those services in the service discovery, they are managed by service discovery.

We have HTTP client factory registrations being loaded from configuration, and those, as you know, can come from anywhere.

If this is an Aspire limitation, it should not reflect on the resources being used on Aspire.

One option would be have an options to override/define the reference name in `WithReference`.

**Reply by davidfowl** [reply]
I think this means we should allow resources to override the name for service discovery.

**Reply by paulomorgado** [reply]
And for `WithReference`.

Let Aspire empower the developer without requiring changes in the resources.

And that would also reflect in the manifest.

**Reply by davidfowl** [reply]
We can restrict the names of the resources to whatever. Those are logical names. I think having an override for the service discovery name is a good feature to add

**Reply by paulomorgado** [reply]
I agree. I just think that, adding to that, there could be an override when adding the reference.

I could do it myself by adding an environment variable. But that becomes very verbose and not the same experience.

**Reply by davidfowl** [reply]
I think that’s what we’d do similar to what happens with WithReference for adding connection strings. You can override the name.

**Reply by michaelrbrown** [reply]
How would we handle azure cloud SQL Server automatically generated domain registrations such as this: var sql = builder.AddSqlServer("sql-chipin.database.windows.net", password);?  The ".database.windows.net" is automatically assigned by default by Azure.  What am I missing?  If this is for cloud native services how would it not handle this scenario?

**Reply by davidfowl** [reply]
When you use aspire with azure sql the name is generated so that it is domain safe. You never add database.windows.net as the aspire resource name. Those are logical names that are used within the app model:

https://github.com/dotnet/aspire/blob/ced70f63a0be48d4b506e55210d1c9cf92820690/src/Aspire.Hosting.Azure.Sql/AzureSqlExtensions.cs#L24

The logical name is used as input to generating a unique name for the fqdn in the azure resource.

**Reply by thematsu** [reply]
> We can restrict the names of the resources to whatever. Those are logical names. I think having an override for the service discovery name is a good feature to add

Hi @davidfowl, is there any functionality that allows such overrides already? I think that this would be a great feature for existing projects to run under .NET Aspire. 

I’m working on a couple of services, lets say A and B. For development, I used to run them under IIS with modified entries in the etc/hosts file. In my current setup, Service A connects to Service B using the serviceb.dev.com URL, which is defined in the appsettings.json file.
Now, to use these services with .NET Aspire, I need to create a new appsettings.json file for the new setup. Ideally, I’d prefer to keep the existing appsettings.json file as is and use service names (including dots) in .NET Aspire, or have the ability to override the service discovery endpoints.

**Reply by davidfowl** [reply]
Why do you need to change anything to use aspire? Are you saying that you want to integrate with aspire’s service discovery or that you have to?

**Reply by paulomorgado** [reply]
@mateuszjarazny, I don't understand how this is forcing you to use another `appsettings.json`. Can you elaborate on that? Can you provide a sample solution?

**Comment by razvangoga** [comment]
@davidfowl encountered the same exception when trying to move and exiting project from a docker compose setup to Aspire vis-a-vis the SqlServer db name. 

**_XX_SOMENAME_SOMEOTHERNAME_** is a valid DB name in SqlServer

``` csharp
Unhandled exception. System.ArgumentException: Resource name 'XX_SOMENAME_SOMEOTHERNAME' is invalid. Name must contain only ASCII letters, digits, and hyphens. (Parameter 'name')
   at Aspire.Hosting.ApplicationModel.ModelName.ValidateName(String target, String name) in /_/src/Aspire.Hosting/ApplicationModel/ModelName.cs:line 39
   at Aspire.Hosting.ApplicationModel.Resource..ctor(String name) in /_/src/Aspire.Hosting/ApplicationModel/Resource.cs:line 30
   at Aspire.Hosting.ApplicationModel.SqlServerDatabaseResource..ctor(String name, String databaseName, SqlServerServerResource parent) in /_/src/Aspire.Hosting/SqlServer/SqlServerDatabaseResource.cs:line 14
   at Aspire.Hosting.SqlServerBuilderExtensions.AddDatabase(IResourceBuilder`1 builder, String name, String databaseName) in /_/src/Aspire.Hosting/SqlServer/SqlServerBuilderExtensions.cs:line 71
```

**Reply by davidfowl** [reply]
There's an overload of AddDatabase that takes the database name and it shouldn't be getting this validation error:

```C#
var db = builder.AddSqlServer("sql").AddDatabase("db", "XX_SOMENAME_SOMEOTHERNAME");
```

**Reply by razvangoga** [reply]
works thanks - didn't notice the overload - i guess that here then the validation would be done on the 1st prm?

if yes - then maybe point the users to this overload in the exception message.

eg : _Resource name 'XX_SOMENAME_SOMEOTHERNAME' is invalid. Name must contain only ASCII letters, digits, and hyphens. (Parameter 'name'). **If you really really wanna use this name, use the .AddDatabase("db", "XX_SOMENAME_SOMEOTHERNAME") overload**_
